core/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java

```diff
@@ -228,8 +226,10 @@ public int hashCode() {
 
     /**
      * Checks for any conflicts between this field type and other.
+     * If strict is true, all properties must be equal.
+     * Otherwise, only properties which must never change in an index are checked.
      */
-    public void checkCompatibility(MappedFieldType other, List<String> conflicts) {
+    public void checkCompatibility(MappedFieldType other, List<String> conflicts, boolean strict) {
```

Comment:
When I did the following test:

```bash
DELETE test

PUT test 
{
  "mappings": {
    "test1": {
      "properties": {
        "s": {
          "type": "binary"
        }
      }
    }
  }
}

PUT test/test2/_mapping
{
  "test2": {
    "properties": {
      "s": {
        "type": "string"
      }
    }
  }
}
```
I got the following response:
```bash
{
   "error": {
      "root_cause": [
         {
            "type": "illegal_argument_exception",
            "reason": "Mapper for [s] conflicts with existing mapping in other types[mapper [s] has different index values]"
         }
      ],
      "type": "illegal_argument_exception",
      "reason": "Mapper for [s] conflicts with existing mapping in other types[mapper [s] has different index values]"
   },
   "status": 400
}
```
I was surprised to see an error about options while the mapped type isn't even the same. I suspect it comes from this method which starts validating options without first comparing the type implementation?